pkg/reconciler/example/iptables.go

```diff
@@ -0,0 +1,297 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/netip"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/reconciler"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/statedb/index"
+)
+
+var (
+	RulePrimaryIndex = statedb.Index[*Rule, *Rule]{
+		Name: "primary",
+		FromObject: func(rule *Rule) index.KeySet {
+			return index.NewKeySet(rule.PrimaryKey())
+		},
+		FromKey: (*Rule).PrimaryKey,
+		Unique:  true,
+	}
+	RuleStatusIndex = reconciler.NewStatusIndex[*Rule]()
+)
+
+var IPTablesCell = cell.Module(
+	"iptables",
+	"IPTables desired state and reconciliation",
+
+	statedb.NewTableCell[*Rule]("iptables-rules", RulePrimaryIndex, RuleStatusIndex),
+
+	cell.ProvidePrivate(
+		func() reconciler.Target[*Rule] { return &iptablesTarget{} },
+		func() reconciler.Config {
+			return reconciler.Config{
+				FullReconcilationInterval: 10 * time.Second,
+				RetryBackoffMinDuration:   time.Second,
+				RetryBackoffMaxDuration:   10 * time.Second,
+			}
+		},
+		func() statedb.Index[*Rule, reconciler.StatusKind] { return RuleStatusIndex },
+	),
+	cell.Invoke(reconciler.Register[*Rule]),
+)
+
+//
+// iptables reconciliation target
+//
+
+type iptablesTarget struct {
+}
+
+// Delete implements reconciler.Target
+func (*iptablesTarget) Delete(_ context.Context, rule *Rule) error {
+	fmt.Printf(">>> iptables %s\n", strings.Join(rule.ToArgs("-D"), " "))
+	return nil
+}
+
+// Sync implements reconciler.Target
+func (*iptablesTarget) Sync(_ context.Context, _ statedb.Iterator[*Rule]) (outOfSync bool, err error) {
+	// TODO:
+	// - "iptables-save" and parse? Or "-C" on each rule?
+	// - Check that cilium chains exist
+	// - Find unexpected rules in cilium chains?
+	return false, nil
+}
+
+// Update implements reconciler.Target
+func (*iptablesTarget) Update(_ context.Context, rule *Rule) error {
+	fmt.Printf(">>> iptables %s\n", strings.Join(rule.ToArgs("-A"), " "))
+	return errors.New("oops iptables error")
+}
+
+func (*iptablesTarget) Init(context.Context) error {
+	// TODO:
+	// - Load kernel modules.
+	// - Probe for features.
+	// - Setup cilium chains.
+	return nil
+}
+
+var _ reconciler.Target[*Rule] = &iptablesTarget{}
+
+//
+// Rule
+//
```

Comment:
Do we need to define all types of "rules" "jump"? Can't the state be the full iptables command? For example, the desired state is `-A POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE` and the realized state is basically a execution of `iptables -A POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE`.

Reply:
I think I shouldn’t have left this very unfinished example in…